libs/@blockprotocol/type-system/rust/src/schema/data_type/constraint/number.rs

```diff
@@ -120,16 +120,110 @@ impl Constraint for NumberSchema {
         other: Self,
     ) -> Result<(Self, Option<Self>), Report<ResolveClosedDataTypeError>> {
         Ok(match (self, other) {
-            (Self::Constrained(lhs), Self::Constrained(rhs)) => {
-                let (combined, remainder) = lhs.combine(rhs)?;
-                (
-                    Self::Constrained(combined),
-                    remainder.map(Self::Constrained),
-                )
+            (Self::Constrained(lhs), Self::Constrained(rhs)) => lhs
+                .combine(rhs)
+                .map(|(lhs, rhs)| (Self::Constrained(lhs), rhs.map(Self::Constrained)))?,
+            (Self::Const { r#const }, Self::Constrained(constraints))
+            | (Self::Constrained(constraints), Self::Const { r#const }) => {
+                constraints.validate_value(&r#const).change_context(
+                    ResolveClosedDataTypeError::UnsatisfiedConstraint(
+                        json!(r#const),
+                        ValueConstraints::Typed(SingleValueConstraints::Number(Self::Constrained(
+                            constraints,
+                        ))),
+                    ),
+                )?;
+
+                (Self::Const { r#const }, None)
+            }
+            (Self::Enum { r#enum }, Self::Constrained(constraints))
+            | (Self::Constrained(constraints), Self::Enum { r#enum }) => {
+                // We use the fast way to filter the values that pass the constraints and collect
+                // them. In most cases this will result in at least one value
+                // passing the constraints.
+                let passed = r#enum
+                    .iter()
+                    .filter(|&value| constraints.is_valid(value))
+                    .copied()
+                    .collect::<Vec<_>>();
+
+                match passed[..] {
+                    [] => {
+                        // We now properly capture errors to return it to the caller.
+                        let _: Vec<()> = r#enum
```

Comment:
`()` implements `FromIterator<()>`, therefore we shouldn't even need the vec here.

Reply:
Oh, good to know, thanks!